<?xml version="1.0"?>
<robot name="Ned" xmlns:xacro="http://ros.org/wiki/xacro">
  <link name="base_link">
    <visual>
      <origin xyz="0.09 0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.09 0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.09 0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.09 0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.09 -0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.09 -0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.09 -0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.09 -0.09 0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.006"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.09 0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.09 0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.09 0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.09 0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.09 -0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.09 -0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.09 -0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.09 -0.09 0.008" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.012" length="0.004"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0.012" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.2 0.004"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.012" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.2 0.004"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 -0.0455 0.052" rpy="0 0 0"/>
      <geometry>
        <box size="0.136 0.091 0.076"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 -0.0455 0.052" rpy="0 0 0"/>
      <geometry>
        <box size="0.136 0.091 0.076"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0.04425 0.052" rpy="0 0 0"/>
      <geometry>
        <box size="0.099 0.0885 0.076"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0.04425 0.052" rpy="0 0 0"/>
      <geometry>
        <box size="0.099 0.0885 0.076"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.0485 0.0485 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.04" length="0.076"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.0485 0.0485 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.04" length="0.076"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.0485 0.0485 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.04" length="0.076"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0485 0.0485 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.04" length="0.076"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.068 -0.071 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.076"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.068 -0.071 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.076"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.068 -0.071 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.076"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.068 -0.071 0.052" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.076"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 -0.0095 0.052" rpy="0 0 0"/>
      <geometry>
        <box size="0.176 0.1195 0.076"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 -0.0095 0.052" rpy="0 0 0"/>
      <geometry>
        <box size="0.176 0.1195 0.076"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0.0975" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.074" length="0.015"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.0975" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.074" length="0.015"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_1" type="revolute">
    <parent link="base_link"/>
    <child link="shoulder_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" lower="-2.8" upper="2.8" velocity="10"/>
    <origin xyz="0 0 0.1065" rpy="0 0 0"/>
  </joint>
  <link name="shoulder_link">
    <visual>
      <origin xyz="0 0 0.111566" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.023" length="0.01"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.111566" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.023" length="0.01"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0.065" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.047" length="0.1135"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.065" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.047" length="0.1135"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.041 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.0278 0.091 0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.041 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.0278 0.091 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.041 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.0278 0.091 0.05"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.041 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.0278 0.091 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.028996 -0.076515 0.028121" rpy="0 0 0"/>
      <geometry>
        <box size="0.085 0.042 0.042"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.028996 -0.076515 0.028121" rpy="0 0 0"/>
      <geometry>
        <box size="0.085 0.042 0.042"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.028993 -0.078529 0.02697" rpy="0 0 0"/>
      <geometry>
        <box size="0.03 0.061 0.052"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.028993 -0.078529 0.02697" rpy="0 0 0"/>
      <geometry>
        <box size="0.03 0.061 0.052"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.077134 -0.076515 0.028121" rpy="0 0 0"/>
      <geometry>
        <box size="0.011 0.03 0.03"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.077134 -0.076515 0.028121" rpy="0 0 0"/>
      <geometry>
        <box size="0.011 0.03 0.03"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_2" type="revolute">
    <parent link="shoulder_link"/>
    <child link="arm_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" lower="-0.8" upper="0.8" velocity="10"/>
    <origin xyz="0 0 0.065" rpy="0 -1.57 0"/>
  </joint>
  <link name="arm_link">
    <visual>
      <origin xyz="0.16 0 0.060549" rpy="0 0 0"/>
      <geometry>
        <box size="0.042 0.04 0.02"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.16 0 0.060549" rpy="0 0 0"/>
      <geometry>
        <box size="0.042 0.04 0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.16 0 0.075315" rpy="0 0 0"/>
      <geometry>
        <box size="0.03 0.03 0.01"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.16 0 0.075315" rpy="0 0 0"/>
      <geometry>
        <box size="0.03 0.03 0.01"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.133834 0.000241 0.058238" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.02 0.012"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.133834 0.000241 0.058238" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.02 0.012"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.08 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.02" length="0.07"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.08 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.02" length="0.07"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.221923 0.011892 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.043" length="0.0986"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.221923 0.011892 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.043" length="0.0986"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.11 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.025" length="0.01"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.11 0 0" rpy="3.141593 1.570793 3.141593"/>
      <geometry>
        <cylinder radius="0.025" length="0.01"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.15 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.07 0.065 0.1"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.15 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.07 0.065 0.1"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_3" type="revolute">
    <parent link="arm_link"/>
    <child link="elbow_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" lower="-1.5" upper="1.3" velocity="10"/>
    <origin xyz="0.221 0.012 0" rpy="0 0 0"/>
  </joint>
  <link name="elbow_link">
    <visual>
      <origin xyz="0.057878 -0.008958 -0.000023" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.035 0.028"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.057878 -0.008958 -0.000023" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.035 0.028"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.032172 0.039196 -0.000023" rpy="0 0 0"/>
      <geometry>
        <box size="0.055 0.07 0.061"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.032172 0.039196 -0.000023" rpy="0 0 0"/>
      <geometry>
        <box size="0.055 0.07 0.061"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.032406 0.08 0" rpy="1.5708 0 0"/>
      <geometry>
        <cylinder radius="0.026" length="0.01"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.032406 0.08 0" rpy="1.5708 0 0"/>
      <geometry>
        <cylinder radius="0.026" length="0.01"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_4" type="revolute">
    <parent link="elbow_link"/>
    <child link="forearm_link"/>
    <axis xyz="-0.000001 0 1"/>
    <limit effort="10000" lower="-2" upper="2" velocity="10"/>
    <origin xyz="0.032 0.042 0" rpy="-1.568653 -1.57 -0.001347"/>
  </joint>
  <link name="forearm_link">
    <visual>
      <origin xyz="0 0 0.11796" rpy="0 0 0"/>
      <geometry>
        <box size="0.057 0.046 0.026"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.11796" rpy="0 0 0"/>
      <geometry>
        <box size="0.057 0.046 0.026"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0.024933 0 0.165857" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.05 0.09"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.024933 0 0.165857" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.05 0.09"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.024933 0 0.165857" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.05 0.09"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.024933 0 0.165857" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.05 0.09"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0.1" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.025" length="0.01"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.1" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.025" length="0.01"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0.08" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.07"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.08" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.02" length="0.07"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_5" type="revolute">
    <parent link="forearm_link"/>
    <child link="wrist_link"/>
    <axis xyz="0.000001 0 1"/>
    <limit effort="10000" lower="-1.5" upper="1.5" velocity="10"/>
    <origin xyz="0 0 0.188" rpy="-1.570187 -1.57 1.570187"/>
  </joint>
  <link name="wrist_link">
    <visual>
      <origin xyz="-0.019278 -0.000743 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.06 0.035 0.031"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.019278 -0.000743 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.06 0.035 0.031"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.009676 0.000228 0.045144" rpy="0 0 0"/>
      <geometry>
        <box size="0.047 0.03 0.03"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.009676 0.000228 0.045144" rpy="0 0 0"/>
      <geometry>
        <box size="0.047 0.03 0.03"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint_6" type="revolute">
    <parent link="wrist_link"/>
    <child link="hand_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" lower="-2.5" upper="2.5" velocity="10"/>
    <origin xyz="0.0197 0.00925 0" rpy="0 1.57 0"/>
  </joint>
  <link name="hand_link">
    <visual>
      <origin xyz="0 0 0.012597" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.0155" length="0.025"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.012597" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.0155" length="0.025"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 -0.001" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.0083" length="0.001"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 -0.001" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.0083" length="0.001"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 -0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.0066" length="0.002"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 -0.003" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.0066" length="0.002"/>
      </geometry>
    </collision>
  </link>
  <link name="base_gripper">
    <visual>
      <origin xyz="0.013321 -0.023019 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.003 0.033 0.06"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.013321 -0.023019 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.003 0.033 0.06"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-0.000465 -0.025439 0.004496" rpy="0 0 0"/>
      <geometry>
        <box size="0.024 0.025 0.034"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.000465 -0.025439 0.004496" rpy="0 0 0"/>
      <geometry>
        <box size="0.024 0.025 0.034"/>
      </geometry>
    </collision>
  </link>
  <joint name="hand_link_base_gripper_joint" type="fixed">
    <parent link="hand_link"/>
    <child link="base_gripper"/>
    <origin xyz="0 0 0.0215" rpy="-1.57 -0.000001 1.57"/>
  </joint>
  <joint name="gripper::left" type="prismatic">
    <parent link="base_gripper"/>
    <child link="jaw_2"/>
    <axis xyz="0 -1 0"/>
    <origin xyz="0.01455 -0.0244 -0.011066" rpy="1.5708 0 0"/>
    <limit effort="1" lower="-0.01" upper="0.01" velocity="5"/>
  </joint>
  <link name="jaw_2">
    <visual>
      <origin xyz="-0.007 -0.0064 0.0386" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.01 0.035"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.007 -0.0064 0.0386" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.01 0.035"/>
      </geometry>
    </collision>
  </link>
  <joint name="gripper::right" type="prismatic">
    <parent link="base_gripper"/>
    <child link="jaw_1"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0.01455 -0.0244 0.009614" rpy="1.5708 0 0"/>
    <limit effort="1" lower="-0.01" upper="0.01" velocity="5"/>
  </joint>
  <link name="jaw_1">
    <visual>
      <origin xyz="-0.007 0.0064 0.0386" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.01 0.035"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.007 0.0064 0.0386" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.01 0.035"/>
      </geometry>
    </collision>
  </link>
</robot>
